### ISSUE TITLE
[HOLD for payment 2023-06-29] [$1000] Console error when selecting a user to assign a task to

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#19772]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat
2. Go to "Assign Task" section
3. Enter the task title and click on the "Next" button
4. Then click on "Confirm task"
5. Click on "To" and select a user
## Expected Result: 
A console error should not appear
## Actual Result: 
Console error when selecting a user to assign a task to

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.19-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/23f3f8f0-c117-423d-a9db-5f4c4be4d0cf

<img width="960" alt="Untitled" src="https://github.com/Expensify/App/assets/43996225/afe3306f-9d14-4c11-affd-d98cd0c33059">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ayazhussain79 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684958586982989

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ca7b76593b61b50d</li>
        <li>Upwork Job ID: 1663406805379948544</li>
        <li>Last Price Increase: 2023-06-13</li>
    </ul>
</details>

**Comment by situchan** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Console error shows after selecting assignee on task assignee selector page

### What is the root cause of that problem?
This came from lack knowledge of lifecycle in functional component.
In `TaskAssigneeSelectorModal`, there's `search_render` measurement.
Timer should start on component mount and should end only on OptionsSelector onLayout callback or component unmount.
In current logic, there are more cases of start/end timer: whenever props or search value are changed because of **`[updateOptions]` dependency**.
https://github.com/Expensify/App/blob/d70b7462cd698a50aa121310f69b815e9f7cc9fa/src/pages/tasks/TaskAssigneeSelectorModal.js#L110-L114
So finally these are call orders in Timing.js:
- end
- start
- end
- start
- getEnvironment promise return
- getEnvironment promise return (error happens here because key is removed by [this line](https://github.com/Expensify/App/blob/d70b7462cd698a50aa121310f69b815e9f7cc9fa/src/libs/actions/Timing.js#L47))
- start
- end
- getEnvironment promise return

We can find out and fix similar issues across the app. (i.e. TaskShareDestinationSelectorModal)

### What changes do you think we should make in order to solve the problem?
call Timing.start only on component mount, Timing.end only on component unmount or OptionsSelector onLayout callback.
To do this, define new useEffect hook with **empty array dependency**:
```
    useEffect(() => {
        Timing.start(CONST.TIMING.SEARCH_RENDER);
        Performance.markStart(CONST.TIMING.SEARCH_RENDER);

        return () => {
            Timing.end(CONST.TIMING.SEARCH_RENDER);
            Performance.markEnd(CONST.TIMING.SEARCH_RENDER);
        };
    }, []);
```

### What alternative solutions did you explore? (Optional)
Early return if key [doesn't](https://github.com/Expensify/App/blob/d70b7462cd698a50aa121310f69b815e9f7cc9fa/src/libs/actions/Timing.js#LL31C4-L31C10) exist [here](https://github.com/Expensify/App/blob/d70b7462cd698a50aa121310f69b815e9f7cc9fa/src/libs/actions/Timing.js#L36)

**Comment by bfitzexpensify** [comment]
Reproduced

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Console error when we select an assignee.

### What is the root cause of that problem?
The error comes from `Timing.end` where we are trying to access `timestampData[eventName]`.
https://github.com/Expensify/App/blob/3ce8e8212c8c9dcc4a7767a1c256555624ad6202/src/libs/actions/Timing.js#L31-L36

Notice that we already have a guard to check if `timestampData[eventName]` exist or not. However, we access it inside the promise, which means something must have deleted it before the promise resolve. We can easily move the check inside the promise, but that doesn't really solve the root cause, but rather just "hiding" the issue. The only way `timestampData[eventName]` is deleted through `Timing.end`, so we can guess that there are one calls to `Timing.end` that already delete the data while the other call's promise just completed.

We have multiple calls of Timing because we call it inside the `useEffect` cleanup function which will be called every time the dependency array changes or the component unmount.
https://github.com/Expensify/App/blob/3ce8e8212c8c9dcc4a7767a1c256555624ad6202/src/pages/tasks/TaskAssigneeSelectorModal.js#L104C22-L114

Notice the dependency array is `updateOptions` which is a function wrapped with `useCallback` that have a dependency array of `props` and `searchValue`. This means, every props changes, we will call Timing.end.

### What changes do you think we should make in order to solve the problem?

As the timing measure is to measure how long the search options to render, I don't think it makes sense to call Timing.end in a `useEffect` cleanup function. We want to measure how long the search to render, not how long the component unmounts or how long any props change. So,

1. Remove Timing and Performance end from `useEffect` cleanup function.
2. Move Timing and Performance start to a new `useEffect` with empty dependency array (on mount).

The above changes are enough to solve this issue, but I see another thing that we can improve, read more below:

<details>
<summary> read more </summary>
After applying my changes above, you will notice that we call `updateOptions` in `useEffect` and `onChangeText`. This means, we are calling `updateOptions` twice every time we type something. Also, calling `updateOptions` in `onChangeText` is wrong because we need to wait for the state changes before calling `updateOptions` (like a `setState` callback). What we need to do instead is replace `updateOptions` to a plain `useEffect` that will run every time the props.reports, personalDetails, betas, or searchValue are changed.

This will improve the code and prevent unnecessary search logic on every props changes.

Let me know if we agree to do this.
</details>

As TaskAssigneeSelectorModal and TaskShareDestinationSelectorModal has a same Timing (and search) logic, we can apply the suggestion on both pages.

**Comment by mallenexpensify** [comment]
Since this already is External and has proposals, I put the below on hold pending this issue since it appears to be a dupe 
- https://github.com/Expensify/App/issues/19462#

**Comment by bfitzexpensify** [comment]
Hey @greg-schroeder - I'm ooo this week so assigning a BZ buddy. Current status: we have a couple of proposals for @eVoloshchak to review. Don't expect you'll need to do much here before I get back next week but appreciate you keeping an eye on it.

**Comment by greg-schroeder** [comment]
I gotchu @bfitzexpensify! Proposals are linked below:

1. https://github.com/Expensify/App/issues/19772#issuecomment-1567504829
2. https://github.com/Expensify/App/issues/19772#issuecomment-1567795676

cc @eVoloshchak

**Comment by eVoloshchak** [comment]
I think we should proceed with @situchan's [proposal](https://github.com/Expensify/App/issues/19772#issuecomment-1567504829). This is a simple and straightforward fix, we indeed just calling a cleanup function multiple times, while should only be calling once

🎀👀🎀 C+ reviewed!
cc: @arosiclair

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
For a given task, the following console error appears when a user is selected and attempted to be assigned to:
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'startTime')

### What is the root cause of that problem?
The root cause of the problem is that the check for the `timestampData` value for a key `eventName` during end performance timing is done before the promise to fetch the Environment returns

The code reference is [here](https://github.com/Expensify/App/blob/main/src/libs/actions/Timing.js#LL31C1-L36C73)

### What changes do you think we should make in order to solve the problem?

To fix the issue, the check for valid value of timestamp for the given `eventName` must be performed when the promise returns:

```
function end(eventName, secondaryName = '') {
    Environment.getEnvironment().then((envName) => {
        if (!timestampData[eventName]) {
            return;
        }
```
        
### What alternative solutions did you explore? (Optional)

**Comment by rojiphil** [comment]
@eVoloshchak @arosiclair Please also have a look at [this proposal](https://github.com/Expensify/App/issues/19772#issuecomment-1579268242) too.

**Comment by greg-schroeder** [comment]
@arosiclair friendly bump on this one!

**Comment by situchan** [comment]
@rojiphil your solution is my alternative solution which is workaround. It's not acceptable since it doesn't fix the root cause.

**Comment by rojiphil** [comment]
> @rojiphil your solution is my alternative solution which is workaround. It's not acceptable since it doesn't fix the root cause.

@situchan, IMHO,

a) My solution is not an alternative solution of yours as it is not in your proposed changes. But I do agree that it was an alternative that you tried and decided not to pursue and propose. I think, it was a workaround for you and, so, you did not propose.

b) I do not think that my [proposed changes](https://github.com/Expensify/App/issues/19772#issuecomment-1579268242) is a workaround. To me, it does not make any sense to check if `timestampData[eventName]` is valid before a promise function returns and, later, use it heavily after the promise returns. That is faulty. So, to me, that is the root cause of this issue. 

Further, in this issue, I do not think that the performance timing feature itself has a bug that mandates us to change the way it is currently implemented. In fact, that may introduce bugs. 

Anyways, I leave it to the discretion of C+ contributors to decide.

**Comment by situchan** [comment]
@rojiphil FYI: if you check other issues, there's chance of alternative solution to be accepted. Why do you think alternative solution cannot be one of proposed changes?
Btw, in this GH, the main root cause is on TaskAssigneeSelectorModal.js so the fix should also come from there.

**Comment by eVoloshchak** [comment]
> Please also have a look at https://github.com/Expensify/App/issues/19772#issuecomment-1579268242 too.

@rojiphil, I think that's the same approach @situchan has proposed [here](https://github.com/Expensify/App/issues/19772#issuecomment-1567504829)
While this would fix the issue, I think it's better if we remove the wrong call to a function instead of returning early if timestamp is valid

**Comment by arosiclair** [comment]
I think @bernhardoj is actually closest to the mark here though doesn't seem completely correct. Effects only run _after_ the component has rendered so we shouldn't start the timer in the effect. For example, @situchan's proposal is measuring how long the component is mounted not how long it takes to render. Also we have another call to end the timer `onLayout` [here](https://github.com/Expensify/App/blob/e20dd3f12dd8b94d50526f388d5a0492286d18d8/src/pages/tasks/TaskAssigneeSelectorModal.js#L208-L211) which I'm not sure is better/worse/same as stopping the timer using an effect.

@bernhardoj can you update your proposal to start the timer at the beginning of the render (top of the function) and do some testing to see if stopping the timer using `onLayout` or `useEffect` is better?

**Comment by situchan** [comment]
Stop timer on onLayout is required to measure until page is fully loaded.
And also need to stop timer in unmount. There might be case of page closed before onLayout is called on slow devices with high-traffic account.

> @situchan's proposal is measuring how long the component is mounted not how long it takes to render

This is already logic happening on main. This issue is just about console error so can be out of scope.
The original PR author added it in component mount because there's no method available in functional component equivalent to constructor in class component.
If we need to fix measure logic as well, we can use useRef to call start timer at the top of function **only once**.

**Comment by situchan** [comment]
I even found out the case of starting timer in componentDidmount for class component:

https://github.com/Expensify/App/blob/fda1a9a8560c8b10de2739fd3475d1988c564a96/src/pages/home/sidebar/SidebarScreen/BaseSidebarScreen.js#L20-L30

So if we need to fix measure logic, should be fixed on all pages across the app. I think this can be definitely separate issue.

**Comment by bernhardoj** [comment]
Replying to @arosiclair,

Updated. Stopping the timing at onLayout will result in longer time, which I think is better

useEffect: ~100
onLayoutL ~300

I agree with @situchan that it's currently inconsistent on how we measure it. So, I don't know whether we should handle all of them in a separate issue or we can just handle this one here. I'll leave it to @arosiclair

**Comment by arosiclair** [comment]
The timer in the `BaseSidebarScreen` is a bit different as it's measuring time across a couple different navigation components. I'm not sure anything is flawed with that so we can leave it as is for now. Lets just focus on `TaskAssigneeSelectorModal` and `TaskShareDestinationSelectorModal `.

Atm it seems the root cause of this issue is that we're timing the render incorrectly, so fixing that is certainly in scope. With that said, it sounds like `onLayout` occurs _after_ mount so maybe this timer is supposed to be measuring layout time not render time?

@cdanwards you authored this in https://github.com/Expensify/App/pull/17992. Can you help clarify what we're trying to time in `TaskAssigneeSelectorModal` and `TaskShareDestinationSelectorModal `? Atm it looks like we're incorrectly timing the render. Does that sound right?

**Comment by situchan** [comment]
I believe that measurement can be completely removed from `TaskAssigneeSelectorModal` and `TaskShareDestinationSelectorModal`.
I think the author just cloned `SearchPage` component and forgot to remove timing part or at least to rename measurement name:
https://github.com/Expensify/App/blob/d91fe14613a9d2f2224296350c8a1d5aee6b3174/src/pages/tasks/TaskAssigneeSelectorModal.js#L109

`SEARCH_RENDER` is for Search page. No reason for it being on task pages.

And on Search page, measurement starts in constructor, ends in onLayout.
As there's no method in function component equivalent to constructor in class component, author just put it in useEffect(() => {}, [])

**Comment by bfitzexpensify** [comment]
hey @cdanwards, just making sure you saw https://github.com/Expensify/App/issues/19772#issuecomment-1584894483

**Comment by arosiclair** [comment]
Pinged Dan in slack

**Comment by cdanwards** [comment]
You are correct, @situchan. I was given the `SearchPage` as a template for this Modal. I wasn't aware that the `CONST.TIMING.SEARCH_RENDER` and the `Timing` were specific to the SearchPage.

**Comment by situchan** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/19772#issuecomment-1567504829)

**Comment by arosiclair** [comment]
Looks good let's push those changes

**Comment by situchan** [comment]
@arosiclair @eVoloshchak I suggest to hold this issue a bit as we cannot test assign task flow without fixing https://github.com/Expensify/App/issues/20807. So I cannot add screenshots.
The [PR](https://github.com/Expensify/App/pull/20808) to fix that issue is being actively reviewed. I keep watching that and will post here after merged.
Btw, here's draft PR for code review: https://github.com/Expensify/App/pull/20840

**Comment by situchan** [comment]
Thanks. That PR is merged but now not able to login (https://expensify.slack.com/archives/C01GTK53T8Q/p1686889937230319) so it's another blocker for testing.

**Comment by situchan** [comment]
@eVoloshchak [PR](https://github.com/Expensify/App/pull/20840) is ready

**Comment by bfitzexpensify** [comment]
Just noting here that I'm going to consider this eligible for the 3-day merging bonus, since it missed by just a few hours and was a little hamstrung with testing because of the log-in issues.

**Comment by bfitzexpensify** [comment]
Offers sent out @ayazhussain79 @situchan @eVoloshchak

**Comment by eVoloshchak** [comment]
@bfitzexpensify, could you hold my payment till tomorrow, please?
I'll try to request the payment using NewDot instead of Upwork, need some additional setting up to do

**Comment by bfitzexpensify** [comment]
Sure!

**Comment by eVoloshchak** [comment]
Not overdue, waiting for the first payment via NewDot to clear, after that will request a payment for this

**Comment by eVoloshchak** [comment]
This has been paid out🎉

**Comment by bfitzexpensify** [comment]
Awesome! Let's close this out.
